### ISSUE TITLE
[C++] Build failure with clang 10.0 and clang-tidy 10.0

### ISSUE THREAD

**Issue by pitrou** [apache/arrow-cookbook#139]
I'm not sure whether this is expected or not. I'm simply trying `make cpp`

```
/home/antoine/arrow/cookbook/cpp/code/basic_arrow.cc:23:1: error: class 'BasicArrow_ReturnNotOkNoMacro_Test' defines a copy constructor and a copy assignment operator but does not define a destructor, a move constructor or a move assignment operator [hicpp-special-member-functions,-warnings-as-errors]
TEST(BasicArrow, ReturnNotOkNoMacro) {
^
/home/antoine/miniconda3/envs/pyarrow/include/gtest/gtest.h:2338:42: note: expanded from macro 'TEST'
#define TEST(test_suite_name, test_name) GTEST_TEST(test_suite_name, test_name)
                                         ^
/home/antoine/miniconda3/envs/pyarrow/include/gtest/gtest.h:2332:3: note: expanded from macro 'GTEST_TEST'
  GTEST_TEST_(test_suite_name, test_name, ::testing::Test, \
  ^
/home/antoine/miniconda3/envs/pyarrow/include/gtest/internal/gtest-internal.h:1355:9: note: expanded from macro 'GTEST_TEST_'
  class GTEST_TEST_CLASS_NAME_(test_suite_name, test_name)                    \
        ^
/home/antoine/miniconda3/envs/pyarrow/include/gtest/internal/gtest-internal.h:1347:3: note: expanded from macro 'GTEST_TEST_CLASS_NAME_'
  test_suite_name##_##test_name##_Test
  ^
note: expanded from here
/home/antoine/arrow/cookbook/cpp/code/basic_arrow.cc:46:1: error: class 'BasicArrow_ReturnNotOk_Test' defines a copy constructor and a copy assignment operator but does not define a destructor, a move constructor or a move assignment operator [hicpp-special-member-functions,-warnings-as-errors]
TEST(BasicArrow, ReturnNotOk) {
^
/home/antoine/miniconda3/envs/pyarrow/include/gtest/gtest.h:2338:42: note: expanded from macro 'TEST'
#define TEST(test_suite_name, test_name) GTEST_TEST(test_suite_name, test_name)
                                         ^
/home/antoine/miniconda3/envs/pyarrow/include/gtest/gtest.h:2332:3: note: expanded from macro 'GTEST_TEST'
  GTEST_TEST_(test_suite_name, test_name, ::testing::Test, \
  ^
/home/antoine/miniconda3/envs/pyarrow/include/gtest/internal/gtest-internal.h:1355:9: note: expanded from macro 'GTEST_TEST_'
  class GTEST_TEST_CLASS_NAME_(test_suite_name, test_name)                    \
        ^
/home/antoine/miniconda3/envs/pyarrow/include/gtest/internal/gtest-internal.h:1347:3: note: expanded from macro 'GTEST_TEST_CLASS_NAME_'
  test_suite_name##_##test_name##_Test
  ^
note: expanded from here
19395 warnings generated.
Suppressed 19481 warnings (19393 in non-user code, 88 NOLINT).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
2 warnings treated as errors
make[3]: *** [CMakeFiles/basic_arrow.dir/build.make:76 : CMakeFiles/basic_arrow.dir/basic_arrow.cc.o] Erreur 2
make[3]: *** Attente des tâches non terminées....
/home/antoine/arrow/cookbook/cpp/code/creating_arrow_objects.cc:23:1: error: class 'CreatingArrowObjects_CreateArrays_Test' defines a copy constructor and a copy assignment operator but does not define a destructor, a move constructor or a move assignment operator [hicpp-special-member-functions,-warnings-as-errors]
TEST(CreatingArrowObjects, CreateArrays) {
^
/home/antoine/miniconda3/envs/pyarrow/include/gtest/gtest.h:2338:42: note: expanded from macro 'TEST'
#define TEST(test_suite_name, test_name) GTEST_TEST(test_suite_name, test_name)
                                         ^
/home/antoine/miniconda3/envs/pyarrow/include/gtest/gtest.h:2332:3: note: expanded from macro 'GTEST_TEST'
  GTEST_TEST_(test_suite_name, test_name, ::testing::Test, \
  ^
/home/antoine/miniconda3/envs/pyarrow/include/gtest/internal/gtest-internal.h:1355:9: note: expanded from macro 'GTEST_TEST_'
  class GTEST_TEST_CLASS_NAME_(test_suite_name, test_name)                    \
        ^
/home/antoine/miniconda3/envs/pyarrow/include/gtest/internal/gtest-internal.h:1347:3: note: expanded from macro 'GTEST_TEST_CLASS_NAME_'
  test_suite_name##_##test_name##_Test
  ^
note: expanded from here
19622 warnings generated.
Suppressed 19709 warnings (19621 in non-user code, 88 NOLINT).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
1 warning treated as error
make[3]: *** [CMakeFiles/creating_arrow_objects.dir/build.make:76 : CMakeFiles/creating_arrow_objects.dir/creating_arrow_objects.cc.o] Erreur 1
make[3]: *** Attente des tâches non terminées....
/home/antoine/arrow/cookbook/cpp/code/datasets.cc:78:1: error: class 'DatasetReadingTest_DatasetRead_Test' defines a copy constructor and a copy assignment operator but does not define a destructor, a move constructor or a move assignment operator [hicpp-special-member-functions,-warnings-as-errors]
TEST_F(DatasetReadingTest, DatasetRead) {
^
/home/antoine/miniconda3/envs/pyarrow/include/gtest/gtest.h:2369:3: note: expanded from macro 'TEST_F'
  GTEST_TEST_(test_fixture, test_name, test_fixture, \
  ^
/home/antoine/miniconda3/envs/pyarrow/include/gtest/internal/gtest-internal.h:1355:9: note: expanded from macro 'GTEST_TEST_'
  class GTEST_TEST_CLASS_NAME_(test_suite_name, test_name)                    \
        ^
/home/antoine/miniconda3/envs/pyarrow/include/gtest/internal/gtest-internal.h:1347:3: note: expanded from macro 'GTEST_TEST_CLASS_NAME_'
  test_suite_name##_##test_name##_Test
  ^
note: expanded from here
/home/antoine/arrow/cookbook/cpp/code/datasets.cc:148:29: error: no member named 'UseAsync' in 'arrow::dataset::ScannerBuilder' [clang-diagnostic-error]
  ASSERT_OK(scanner_builder.UseAsync(true));
                            ^
/home/antoine/arrow/cookbook/cpp/code/common.h:33:73: note: expanded from macro 'ASSERT_OK'
  for (const ::arrow::Status& _st = ::arrow::internal::GenericToStatus((expr)); \
                                                                        ^
31717 warnings and 1 error generated.
Error while processing /home/antoine/arrow/cookbook/cpp/code/datasets.cc.
Suppressed 31851 warnings (31716 in non-user code, 135 NOLINT).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
1 warning treated as error
```

**Comment by pitrou** [comment]
cc @westonpace

**Comment by westonpace** [comment]
It's not expected.  When I originally put this together I was able to build with no warnings.  Since then Arrow-C++ has changed to clang 12.  The build server for the cookbook is using clang 11.  The googletest version should be pinned.

A less fragile fix perhaps would be to use a much less strict set of clang-tidy rules.  The current rules are everything minus some explicitly disabled checks.  This is potentially fragile as newer versions of clang-tidy add more rules.

**Comment by westonpace** [comment]
Or it's possible that this is the root cause:

```
note: expanded from here
/home/antoine/arrow/cookbook/cpp/code/datasets.cc:148:29: error: no member named 'UseAsync' in 'arrow::dataset::ScannerBuilder' [clang-diagnostic-error]
  ASSERT_OK(scanner_builder.UseAsync(true));
```

The `UseAsync` method went away fairly recently so this recipe needs fixed.  I'll add this to my list of things to look at Friday.

**Comment by westonpace** [comment]
I resolved `UseAsync` so that error should go away.

The googletest errors I can reproduce if I use googletest 1.10 instead of 1.11.  The googletest dependency is defined here: https://github.com/apache/arrow-cookbook/blob/main/cpp/code/CMakeLists.txt#L9-L18

I also noticed that your log files show it is using include headers from a conda build of googletest: `/home/antoine/miniconda3/envs/pyarrow/include/gtest/gtest.h`

I'm not sure if the call `include(GoogleTest)` is somehow finding the conda googletest instead of the one we just built or if the include path is getting specified in the wrong order.  I suspect it is the latter.  I do notice on my system that the compile command is:

```
/home/pace/miniconda3/envs/conbench3/bin/cmake -E __run_co_compile --tidy="clang-tidy-12;--extra-arg-before=--driver-mode=g++" --source=/home/pace/dev/arrow-cookbook/cpp/code/datasets.cc -- /usr/bin/clang++-12  -isystem /home/pace/dev/arrow-cookbook/cpp/code/conbench-build/_deps/googletest-src/googletest/include -isystem /home/pace/dev/arrow-cookbook/cpp/code/conbench-build/_deps/googletest-src/googletest -fvisibility-inlines-hidden -std=c++17 -fmessage-length=0 -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /home/pace/miniconda3/envs/conbench3/include -stdlib=libstdc++ -g -Wall -Wextra -Wpedantic -Werror -std=gnu++17 -MD -MT CMakeFiles/datasets.dir/datasets.cc.o -MF CMakeFiles/datasets.dir/datasets.cc.o.d -o CMakeFiles/datasets.dir/datasets.cc.o -c /home/pace/dev/arrow-cookbook/cpp/code/datasets.cc
```

This adds the include path `/home/pace/miniconda3/envs/conbench3/include` at the end of the includes list (after the include for google test) which would explain your behavior.  I can't figure out where in CMakeLists.txt it is specifying that.  Do you have any idea?  If not I can keep messing around with cmake.

**Comment by pitrou** [comment]
Is it actually required to use googletest 1.11?

**Comment by pitrou** [comment]
Note that I'm using a conda environment because for some reason this requires CMake 3.19, which is... extremely recent.

**Comment by westonpace** [comment]
I use a conda environment for the same reasons.  I don't really have the cmake skills or the time at the moment to support too many different build configurations.  There is an environment.yml file in the cpp directory which you can use to create a fresh conda environment if that is easier.  I find cmake/build stuff rather tedious and since this project doesn't have to be super portable I'd rather not support a huge range of configurations.

The project's conda environment doesn't have googletest installed.  In theory it shouldn't be a problem if it is installed.  CMake should be able to prioritize project-built 3rd party includes above conda 3rd party includes.  In practice, I can't figure out how to do that.

Also, ideally, I'd be able to configure clang-tidy to ignore 3rd party code.  It can do this but it doesn't work for macros, which is why googletest was giving me trouble.

I've created a PR that drops the googletest version to 1.10: https://github.com/apache/arrow-cookbook/pull/144
It only required adding one more rule to our clang-tidy ignore list.

**Comment by westonpace** [comment]
Also, the cookbook GHA build currently runs against Arrow from conda, which means that we can't use the latest features in recipes.  This is not ideal.  Using a nightly build of arrow-cpp would probably be much better.

**Comment by pitrou** [comment]
Should be fixed by #144 144
